<a href="https://colab.research.google.com/github/yunhyeongil/pandas_base/blob/main/%EB%A1%9C%EC%9D%B4%ED%84%B0_%EB%89%B4%EC%8A%A4_%EB%B6%84%EB%A5%98%ED%95%98%EA%B8%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##  딥 러닝을 이용한 자연어 처리 입문
- LSTM으로 로이터 뉴스 분류하기 일단 따라해보기
https://wikidocs.net/22933

- 테스트 정확도: 0.7177

In [ ]:
from tensorflow.keras.datasets import reuters
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [ ]:
from tensorflow.keras.datasets import reuters
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM, Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
(X_train, y_train), (X_test, y_test) = reuters.load_data(num_words=1000, test_split=0.2)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
max_len = 100
X_train = pad_sequences(X_train, maxlen=max_len) # 훈련용 뉴스 기사 패딩
X_test = pad_sequences(X_test, maxlen=max_len) # 테스트용 뉴스 기사 패딩

In [ ]:
y_train = to_categorical(y_train) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
model = Sequential()
model.add(Embedding(1000, 120))
model.add(LSTM(120))
model.add(Dense(46, activation='softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

In [ ]:
history = model.fit(X_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(X_test, y_test))

Epoch 1/30
71/71 [==============================] - 29s 376ms/step - loss: 2.5736 - acc: 0.3671 - val_loss: 2.1150 - val_acc: 0.4733

Epoch 00001: val_acc improved from -inf to 0.47329, saving model to best_model.h5
Epoch 2/30
71/71 [==============================] - 27s 374ms/step - loss: 2.0541 - acc: 0.4856 - val_loss: 2.0032 - val_acc: 0.5045

Epoch 00002: val_acc improved from 0.47329 to 0.50445, saving model to best_model.h5
Epoch 3/30
71/71 [==============================] - 26s 370ms/step - loss: 1.8694 - acc: 0.5237 - val_loss: 1.8107 - val_acc: 0.5298

Epoch 00003: val_acc improved from 0.50445 to 0.52983, saving model to best_model.h5
Epoch 4/30
71/71 [==============================] - 26s 370ms/step - loss: 1.7621 - acc: 0.5332 - val_loss: 1.7583 - val_acc: 0.5570

Epoch 00004: val_acc improved from 0.52983 to 0.55699, saving model to best_model.h5
Epoch 5/30
71/71 [==============================] - 26s 367ms/step - loss: 1.6510 - acc: 0.5726 - val_loss: 1.6785 - val_acc: 0

In [ ]:
loaded_model = load_model('best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(X_test, y_test)[1]))

71/71 [==============================] - 3s 31ms/step - loss: 1.2276 - acc: 0.7177

 테스트 정확도: 0.7177


## 연습 1

- 테스트 정확도: 0.7021

In [ ]:
from tensorflow.keras.datasets import reuters
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical

In [ ]:
max_words = 1000
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words, test_split=0.2)

2113536/2110848 [==============================] - 0s 0us/step


/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}')
print(f'테스트 샘플의 수 : {len(x_test)}')
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


**pad_sequences()**

pad_sequences 함수는 시퀀스를 패딩을 하여 리스트의 길이를 동일하게 맞춰준다


```
keras.preprocessing.sequence.pad_sequences(sequences, maxlen=None, dtype='int32', padding='pre', truncating='pre', value=0.0)
```
- sequences: 리스트의 리스트로, 각 성분이 시퀀스이다
- maxlen: 정수, 모든 시퀀스의 최대 길이를 설정하여 제한한다. 10을 넣으면 10보다 큰 것들은 자른다.
- dtype: 출력 시퀀스의 자료형. 가변적 길이의 문자열로 시퀀스를 패딩 하려면object를 사용하시면 됩니다.
- padding: 문자열이 들어간다, 'pre'가 디폴트 값으로 앞쪽에 0이 추가되고, 'post'는 뒤쪽으로 0이 추가되어 각 시퀀스를 패딩 한다.
- truncating: 문자열, 'pre'는 길이가 초과됐을 때 앞쪽을 자르고 'post'는 maxlen보다 큰 시퀀스의 끝의 값들을 제거한다
- value: 부동소수점 혹은 문자열, 패딩 할 값.

리턴 값으로는 Numpy 배열을 리턴한다. 위의 인수들 중에서 필요한 것들만 뽑아 쓰면 된다


In [ ]:
max_len = 100
x_train = pad_sequences(x_train, maxlen=max_len) # 훈련용 뉴스 기사 패딩
x_test = pad_sequences(x_test, maxlen=max_len) # 테스트용 뉴스 기사 패딩

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 100)
x_test shape: (2246, 100)


In [ ]:
print(x_train[:5])
print(x_test[:5])

[[  0   0   0   0   0   0   0   0   0   0   0   0   0   1   2   2   8  43
   10 447   5  25 207 270   5   2 111  16 369 186  90  67   7  89   5  19
  102   6  19 124  15  90  67  84  22 482  26   7  48   4  49   8 864  39
  209 154   6 151   6  83  11  15  22 155  11  15   7  48   9   2   2 504
    6 258   6 272  11  15  22 134  44  11  15  16   8 197   2  90  67  52
   29 209  30  32 132   6 109  15  17  12]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   1   2 699   2   2  56   2   2   9  56
    2   2  81   5   2  57 366 737 132  20   2   7   2  49   2   2   2   2
  699   2   8   7  10 241  16 855 129 231 783   5   4 587   2   2   2 775
    7  48  34 191  44  35   2 505  17  12]
 [ 19 758  15  53 959  47   2  15  14  19 132  15  39 965  32  11  14 147
   72  11 180 183 187  44  11  14 102  19  11 123 186  90  67 960   4  78
   13  68 467 511 110  59 

In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
print(y_train[:5])
print(y_test[:5])

[[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM, Embedding, Dropout, Activation, Bidirectional
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

**임베딩**

```
model = Sequential()
model.add(Embedding(1000, 64, input_length=10))
# 모델은 (batch, input_length)의 크기를 가진 정수 행렬만 인풋으로 전달받습니다.
# 인풋 내 가장 큰 정수(다시 말해 단어 색인)는
# 999(어휘목록 사이즈)보다 커서는 안됩니다.
# 현재 model.output_shape == (None, 10, 64), 여기서 배치 차원은 None입니다.

input_array = np.random.randint(1000, size=(32, 10))

model.compile('rmsprop', 'mse')
output_array = model.predict(input_array)
assert output_array.shape == (32, 10, 64)
```

- input_dim: 정수 > 0. 어휘목록의 크기, 다시 말해, 최대 정수 색인 + 1.
- output_dim: 정수 >= 0. 밀집 임베딩의 차원.
- embeddings_initializer: embeddings 행렬의 초기값 설정기 (초기값 설정기 참조).
- embeddings_regularizer: embeddings 행렬에 적용되는 정규화 함수 (정규화 참조).
- embeddings_constraint: embeddings행렬에 적용되는 제약 함수 (제약 참조).
- mask_zero: 인풋 값 0을 마스크 처리해야 할 특수 "패딩"값으로 다룰 것인지 여부. 순환 레이어를 사용해 가변적 길이의 인풋을 전달받을 때 유용합니다. 이 값이 True인 경우 모델 내 차후 레이어는 마스킹을 지원해야 하며 그렇지 않은 경우 예외가 발생됩니다. mask_zero가 참으로 설정된 경우, 색인 0은 어휘목록에서 사용할 수 없습니다 (input_dim이 vocabulary + 1과 동일한 크기를 가져야 하기 때문입니다).
- input_length: 인풋 시퀀스의 길이로, 그 길이가 불변해야 합니다. 상위 Dense 레이어와 Flatten 레이어를 연결하려면 이 인수가 필요합니다
(그렇지 않고서는 밀집 아웃풋의 형태를 계산할 수 없습니다).

In [ ]:
model = Sequential()
model.add(Embedding(max_words, 128, input_length=max_len))
model.add(Bidirectional(LSTM(64)))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])
# loss='binary_crossentropy'  테스트 정확도: 0.5757

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/30
71/71 [==============================] - 30s 380ms/step - loss: 1.6108 - acc: 0.6022 - val_loss: 1.7092 - val_acc: 0.5797

Epoch 00001: val_acc improved from 0.57569 to 0.57970, saving model to best_model.h5
Epoch 2/30
71/71 [==============================] - 26s 370ms/step - loss: 1.5616 - acc: 0.6100 - val_loss: 1.7039 - val_acc: 0.5766

Epoch 00002: val_acc did not improve from 0.57970
Epoch 3/30
71/71 [==============================] - 27s 377ms/step - loss: 1.5771 - acc: 0.6065 - val_loss: 1.7067 - val_acc: 0.5833

Epoch 00003: val_acc improved from 0.57970 to 0.58326, saving model to best_model.h5
Epoch 4/30
71/71 [==============================] - 27s 374ms/step - loss: 1.5471 - acc: 0.6207 - val_loss: 1.6837 - val_acc: 0.5899

Epoch 00004: val_acc improved from 0.58326 to 0.58994, saving model to best_model.h5
Epoch 5/30
71/71 [==============================] - 26s 365ms/step - loss: 1.4963 - acc: 0.6310 - val_loss: 1.6404 - val_acc: 0.5931

Epoch 00005: val_acc impr

In [ ]:
loaded_model = load_model('best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 3s 29ms/step - loss: 1.2965 - acc: 0.7021

 테스트 정확도: 0.7021


## Tokenizer Sequence To Matrix Binary Mode 

- 테스트 정확도: 0.8130

In [ ]:
from tensorflow.keras.datasets import reuters

In [ ]:
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
word_index = reuters.get_word_index(path="reuters_word_index.json")

557056/550378 [==============================] - 0s 0us/step


In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}') # train sequences
print(f'테스트 샘플의 수 : {len(x_test)}') # test sequences
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


In [ ]:
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(8982,)
(2246,)
(8982,)
(2246,)


In [ ]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


In [ ]:
import keras
from tensorflow.keras.utils import to_categorical
from keras.preprocessing.text import Tokenizer

**Tokenizer란?**

Tokenizer은 사전에 있는 단어의 순서에 맞게 단어를 turning하면서 말뭉치를 벡터화시킨다.


```
tf.keras.preprocessing.text.Tokenizer(
    num_words=None, 
    filters='!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n', 
    lower=True,
    split=' ', 
    char_level=False, 
    oov_token=None, 
    document_count=0, 
    **kwargs
)
```
- num_words : 단어 빈도가 많은 순서로 num_words개의 단어를 보존한다.
- filters : 걸러낼 문자 리스트를 적어준다. 디폴트는 '!"#$%&()*+,-./:;<=>?@[\]^_`{|}~\t\n' 이다.
- lower : 입력받은 문자열을 소문자로 변환할지를 True, False로 적어준다.
- split : 문자열을 적어 줘야 하고, 단어를 분리하는 기준을 적어준다.
- char_level : True인 경우 모든 문자가 토큰으로 처리된다.
- oov_token : 값이 지정된 경우, text_to_sequence 호출 과정에서 word_index에 추가되어 out-of-vocabulary words를 대체한다.


In [ ]:
# Vectorizing sequence data
max_words= 10000
tokenizer= Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode="binary")
x_test = tokenizer.sequences_to_matrix(x_test, mode="binary")

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 10000)
x_test shape: (2246, 10000)


In [ ]:
print(x_train[:5])
print(x_test[:5])

[[0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]
[[0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]


In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
print(y_train[:5])
print(y_test[:5])

[[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
[[0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.

In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
# 모델은 인터넷에서 찾아본것 중 테스트 정확도가 높게 나온걸 사용했다..
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/30
71/71 [==============================] - 21s 86ms/step - loss: 2.2194 - acc: 0.5571 - val_loss: 1.0281 - val_acc: 0.7747

Epoch 00001: val_acc improved from -inf to 0.77471, saving model to best_model.h5
Epoch 2/30
71/71 [==============================] - 6s 82ms/step - loss: 0.7112 - acc: 0.8494 - val_loss: 0.8451 - val_acc: 0.8041

Epoch 00002: val_acc improved from 0.77471 to 0.80410, saving model to best_model.h5
Epoch 3/30
71/71 [==============================] - 6s 81ms/step - loss: 0.3874 - acc: 0.9166 - val_loss: 0.8166 - val_acc: 0.8059

Epoch 00003: val_acc improved from 0.80410 to 0.80588, saving model to best_model.h5
Epoch 4/30
71/71 [==============================] - 5s 75ms/step - loss: 0.2558 - acc: 0.9425 - val_loss: 0.8198 - val_acc: 0.8112

Epoch 00004: val_acc improved from 0.80588 to 0.81122, saving model to best_model.h5
Epoch 5/30
71/71 [==============================] - 5s 77ms/step - loss: 0.1975 - acc: 0.9550 - val_loss: 0.8507 - val_acc: 0.8081

Ep

In [ ]:
loaded_model = load_model('best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 12ms/step - loss: 0.9151 - acc: 0.8130

 테스트 정확도: 0.8130


## Tokenizer Sequence To Matrix Count Mode 

- 테스트 정확도: 0.8139

In [ ]:
from tensorflow.keras.datasets import reuters

In [ ]:
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
word_index = reuters.get_word_index(path="reuters_word_index.json")

In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}') # train sequences
print(f'테스트 샘플의 수 : {len(x_test)}') # test sequences
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


In [ ]:
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(8982,)
(2246,)
(8982,)
(2246,)


In [ ]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


In [ ]:
import keras
from tensorflow.keras.utils import to_categorical
from keras.preprocessing.text import Tokenizer

In [ ]:
# Vectorizing sequence data
max_words= 10000
tokenizer= Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode="count")
x_test = tokenizer.sequences_to_matrix(x_test, mode="count")

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 10000)
x_test shape: (2246, 10000)


In [ ]:
print(x_train[:5])
print(x_test[:5])

[[0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]
[[0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]


In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
# 모델은 인터넷에서 찾아본것 중 테스트 정확도가 높게 나온걸 사용했다..
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('c_best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/30
71/71 [==============================] - 20s 81ms/step - loss: 2.1928 - acc: 0.5714 - val_loss: 1.0372 - val_acc: 0.7903

Epoch 00001: val_acc improved from -inf to 0.79029, saving model to c_best_model.h5
Epoch 2/30
71/71 [==============================] - 5s 76ms/step - loss: 0.7291 - acc: 0.8558 - val_loss: 0.8735 - val_acc: 0.8085

Epoch 00002: val_acc improved from 0.79029 to 0.80855, saving model to c_best_model.h5
Epoch 3/30
71/71 [==============================] - 6s 78ms/step - loss: 0.4143 - acc: 0.9164 - val_loss: 0.8724 - val_acc: 0.8139

Epoch 00003: val_acc improved from 0.80855 to 0.81389, saving model to c_best_model.h5
Epoch 4/30
71/71 [==============================] - 6s 81ms/step - loss: 0.2830 - acc: 0.9431 - val_loss: 0.9044 - val_acc: 0.8117

Epoch 00004: val_acc did not improve from 0.81389
Epoch 5/30
71/71 [==============================] - 5s 77ms/step - loss: 0.2214 - acc: 0.9465 - val_loss: 0.9540 - val_acc: 0.8032

Epoch 00005: val_acc did not im

In [ ]:
loaded_model = load_model('c_best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 11ms/step - loss: 0.8724 - acc: 0.8139

 테스트 정확도: 0.8139


## Tokenizer Sequence To Matrix Frequency Mode 

- 테스트 정확도: 0.8112

In [ ]:
from tensorflow.keras.datasets import reuters

In [ ]:
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
word_index = reuters.get_word_index(path="reuters_word_index.json")

In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}') # train sequences
print(f'테스트 샘플의 수 : {len(x_test)}') # test sequences
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


In [ ]:
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(8982,)
(2246,)
(8982,)
(2246,)


In [ ]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


In [ ]:
import keras
from tensorflow.keras.utils import to_categorical
from keras.preprocessing.text import Tokenizer

In [ ]:
# Vectorizing sequence data
max_words= 10000
tokenizer= Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode="freq")
x_test = tokenizer.sequences_to_matrix(x_test, mode="freq")

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 10000)
x_test shape: (2246, 10000)


In [ ]:
print(x_train[:5])
print(x_test[:5])

[[0.         0.01149425 0.         ... 0.         0.         0.        ]
 [0.         0.01785714 0.         ... 0.         0.         0.        ]
 [0.         0.00719424 0.         ... 0.         0.         0.        ]
 [0.         0.00446429 0.         ... 0.         0.         0.        ]
 [0.         0.00990099 0.         ... 0.         0.         0.        ]]
[[0.         0.00689655 0.         ... 0.         0.         0.        ]
 [0.         0.00134228 0.         ... 0.         0.         0.        ]
 [0.         0.00438596 0.         ... 0.         0.         0.        ]
 [0.         0.00581395 0.         ... 0.         0.         0.        ]
 [0.         0.00534759 0.         ... 0.         0.         0.        ]]


In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
# 모델은 인터넷에서 찾아본것 중 테스트 정확도가 높게 나온걸 사용했다..
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('f_best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/30
71/71 [==============================] - 21s 92ms/step - loss: 3.4152 - acc: 0.4053 - val_loss: 2.2040 - val_acc: 0.3629

Epoch 00001: val_acc improved from -inf to 0.36287, saving model to f_best_model.h5
Epoch 2/30
71/71 [==============================] - 6s 83ms/step - loss: 2.1217 - acc: 0.4579 - val_loss: 1.9182 - val_acc: 0.5151

Epoch 00002: val_acc improved from 0.36287 to 0.51514, saving model to f_best_model.h5
Epoch 3/30
71/71 [==============================] - 5s 76ms/step - loss: 1.8323 - acc: 0.5264 - val_loss: 1.7667 - val_acc: 0.5227

Epoch 00003: val_acc improved from 0.51514 to 0.52271, saving model to f_best_model.h5
Epoch 4/30
71/71 [==============================] - 6s 79ms/step - loss: 1.7119 - acc: 0.5515 - val_loss: 1.6387 - val_acc: 0.6011

Epoch 00004: val_acc improved from 0.52271 to 0.60107, saving model to f_best_model.h5
Epoch 5/30
71/71 [==============================] - 5s 74ms/step - loss: 1.5770 - acc: 0.6272 - val_loss: 1.5301 - val_acc: 0.

In [ ]:
loaded_model = load_model('f_best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 13ms/step - loss: 0.7790 - acc: 0.8112

 테스트 정확도: 0.8112


## Tokenizer Sequence To Matrix TF-IDF Mode

- 테스트 정확도: 0.8206

In [ ]:
from tensorflow.keras.datasets import reuters

In [ ]:
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)

2113536/2110848 [==============================] - 0s 0us/step


/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
word_index = reuters.get_word_index(path="reuters_word_index.json")

557056/550378 [==============================] - 0s 0us/step


In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}') # train sequences
print(f'테스트 샘플의 수 : {len(x_test)}') # test sequences
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


In [ ]:
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(8982,)
(2246,)
(8982,)
(2246,)


In [ ]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


In [ ]:
import keras
import numpy as np
from tensorflow.keras.utils import to_categorical
from keras.preprocessing.text import Tokenizer

In [ ]:
# Vectorizing sequence data
# max_words를 1000 으로 하면 테스트 정확도: 0.8161
# max_words를 50000 으로 하면 테스트 정확도: 0.8201
# max_words를 100000 으로 하면 RAAM사용량 초과로 세션종료됨.

max_words= 10000
tokenizer= Tokenizer(num_words=max_words)
tokenizer.fit_on_sequences(x_train.tolist())
tokenizer.fit_on_sequences(x_test.tolist())
x_train = np.asarray(tokenizer.sequences_to_matrix(x_train.tolist(), mode="tfidf"))
x_test = np.asarray(tokenizer.sequences_to_matrix(x_test.tolist(), mode="tfidf"))

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 10000)
x_test shape: (2246, 10000)


In [ ]:
print(x_train[:5])
print(x_test[:5])

[[0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]]
[[0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]
 [0.         0.69310265 0.         ... 0.         0.         0.        ]]


In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
# 모델은 인터넷에서 찾아본것 중 테스트 정확도가 높게 나온걸 사용했다..
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('t_best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=30, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/30
71/71 [==============================] - 20s 82ms/step - loss: 1.9060 - acc: 0.6067 - val_loss: 0.8934 - val_acc: 0.8206

Epoch 00001: val_acc improved from -inf to 0.82057, saving model to t_best_model.h5
Epoch 2/30
71/71 [==============================] - 5s 74ms/step - loss: 0.4139 - acc: 0.9158 - val_loss: 0.9048 - val_acc: 0.8188

Epoch 00002: val_acc did not improve from 0.82057
Epoch 3/30
71/71 [==============================] - 5s 75ms/step - loss: 0.2540 - acc: 0.9479 - val_loss: 0.9798 - val_acc: 0.8130

Epoch 00003: val_acc did not improve from 0.82057
Epoch 4/30
71/71 [==============================] - 5s 75ms/step - loss: 0.2084 - acc: 0.9551 - val_loss: 1.0609 - val_acc: 0.8112

Epoch 00004: val_acc did not improve from 0.82057
Epoch 5/30
71/71 [==============================] - 5s 74ms/step - loss: 0.1810 - acc: 0.9564 - val_loss: 1.0913 - val_acc: 0.8041

Epoch 00005: val_acc did not improve from 0.82057
Epoch 00005: early stopping


In [ ]:
loaded_model = load_model('t_best_model.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 11ms/step - loss: 0.8934 - acc: 0.8206

 테스트 정확도: 0.8206


## TF-IDF Mode 를 사용한게 테스트 정확도가 0.8206으로 제일 높게 나왔다.(옵티마이저 adam)

## 모델 옵티마이저 변경

## 옵티마이저 사용법
https://keras.io/ko/optimizers/

- SGD 테스트 정확도: 0.8170

- RMSProp 테스트 정확도: 0.8112

- Adamax 테스트 정확도: 0.8175

In [ ]:
from tensorflow.keras.datasets import reuters

In [ ]:
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=None, test_split=0.2)

2113536/2110848 [==============================] - 0s 0us/step


/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:143: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/reuters.py:144: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_test, y_test = np.array(xs[idx:]), np.array(labels[idx:])


In [ ]:
word_index = reuters.get_word_index(path="reuters_word_index.json")

557056/550378 [==============================] - 0s 0us/step


In [ ]:
print(f'훈련 샘플의 수 : {len(x_train)}') # train sequences
print(f'테스트 샘플의 수 : {len(x_test)}') # test sequences
num_classes = max(y_train) + 1
print(f'클래스의 수 : {num_classes}')

훈련 샘플의 수 : 8982
테스트 샘플의 수 : 2246
클래스의 수 : 46


In [ ]:
index_to_word = {}
for key, value in word_index.items():
    index_to_word[value] = key

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(8982,)
(2246,)
(8982,)
(2246,)


In [ ]:
print(' '.join([index_to_word[x] for x in x_train[0]]))
print(y_train[0])

the wattie nondiscriminatory mln loss for plc said at only ended said commonwealth could 1 traders now april 0 a after said from 1985 and from foreign 000 april 0 prices its account year a but in this mln home an states earlier and rise and revs vs 000 its 16 vs 000 a but 3 psbr oils several and shareholders and dividend vs 000 its all 4 vs 000 1 mln agreed largely april 0 are 2 states will billion total and against 000 pct dlrs
3


In [ ]:
import keras
import numpy as np
from tensorflow.keras.utils import to_categorical
from keras.preprocessing.text import Tokenizer

In [ ]:
# Vectorizing sequence data
max_words= 10000
tokenizer= Tokenizer(num_words=max_words)
tokenizer.fit_on_sequences(x_train.tolist())
tokenizer.fit_on_sequences(x_test.tolist())
x_train = np.asarray(tokenizer.sequences_to_matrix(x_train.tolist(), mode="tfidf"))
x_test = np.asarray(tokenizer.sequences_to_matrix(x_test.tolist(), mode="tfidf"))

In [ ]:
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (8982, 10000)
x_test shape: (2246, 10000)


In [ ]:
y_train = to_categorical(y_train, num_classes) # 훈련용 뉴스 기사 레이블의 원-핫 인코딩
y_test = to_categorical(y_test, num_classes) # 테스트용 뉴스 기사 레이블의 원-핫 인코딩

In [ ]:
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.models import load_model

In [ ]:
# del model # 모델 지우기

In [ ]:
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

In [ ]:
es = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=4) # 훈련 조기 종료
mc = ModelCheckpoint('./t_best_model_adamax.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True) # 훈련 과정에서 검증 데이터가 가장 높았을 때 저장

In [ ]:
# model.compile(loss='categorical_crossentropy', optimizer='sgd', metrics=['acc'])  # 확률적 경사 하강법 # t_best_model_sgd.h5
# model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['acc']) # RMSProp # 일반적으로 순환 신경망(Recurrent Neural Networks)의 옵티마이저로 많이 사용
model.compile(loss='categorical_crossentropy', optimizer='adamax', metrics=['acc']) # Adamax # 무한 노름(infinity norm)에 기반한 Adam의 변형

In [ ]:
history = model.fit(x_train, y_train, batch_size=128, epochs=100, callbacks=[es, mc], validation_data=(x_test, y_test))

Epoch 1/100
71/71 [==============================] - 6s 78ms/step - loss: 2.1622 - acc: 0.5526 - val_loss: 1.1194 - val_acc: 0.7716

Epoch 00001: val_acc improved from -inf to 0.77159, saving model to ./t_best_model_adamax.h5
Epoch 2/100
71/71 [==============================] - 5s 75ms/step - loss: 0.7635 - acc: 0.8454 - val_loss: 0.9451 - val_acc: 0.8028

Epoch 00002: val_acc improved from 0.77159 to 0.80276, saving model to ./t_best_model_adamax.h5
Epoch 3/100
71/71 [==============================] - 5s 76ms/step - loss: 0.5098 - acc: 0.8973 - val_loss: 0.8976 - val_acc: 0.8085

Epoch 00003: val_acc improved from 0.80276 to 0.80855, saving model to ./t_best_model_adamax.h5
Epoch 4/100
71/71 [==============================] - 5s 74ms/step - loss: 0.3523 - acc: 0.9321 - val_loss: 0.8761 - val_acc: 0.8085

Epoch 00004: val_acc did not improve from 0.80855
Epoch 5/100
71/71 [==============================] - 5s 74ms/step - loss: 0.2898 - acc: 0.9378 - val_loss: 0.8765 - val_acc: 0.8090



In [ ]:
loaded_model = load_model('t_best_model_sgd.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 11ms/step - loss: 0.8481 - acc: 0.8170

 테스트 정확도: 0.8170


In [ ]:
loaded_model = load_model('t_best_model_rmsprop.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 10ms/step - loss: 1.1061 - acc: 0.8112

 테스트 정확도: 0.8112


In [ ]:
loaded_model = load_model('t_best_model_adamax.h5')
print("\n 테스트 정확도: %.4f" % (loaded_model.evaluate(x_test, y_test)[1]))

71/71 [==============================] - 1s 10ms/step - loss: 0.8805 - acc: 0.8175

 테스트 정확도: 0.8175
